staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go

```diff
@@ -244,15 +245,37 @@ func (cfgCtlr *configController) updateObservations() {
 	}
 }
 
+// used from the unit tests only.
+func (cfgCtlr *configController) getPriorityLevelState(plName string) *priorityLevelState {
+	cfgCtlr.lock.Lock()
+	defer cfgCtlr.lock.Unlock()
+	return cfgCtlr.priorityLevelStates[plName]
+}
+
 func (cfgCtlr *configController) Run(stopCh <-chan struct{}) error {
-	defer cfgCtlr.configQueue.ShutDown()
+	defer utilruntime.HandleCrash()
+
 	klog.Info("Starting API Priority and Fairness config controller")
 	if ok := cache.WaitForCacheSync(stopCh, cfgCtlr.plInformerSynced, cfgCtlr.fsInformerSynced); !ok {
 		return fmt.Errorf("Never achieved initial sync")
 	}
+
 	klog.Info("Running API Priority and Fairness config worker")
-	wait.Until(cfgCtlr.runWorker, time.Second, stopCh)
-	klog.Info("Shutting down API Priority and Fairness config worker")
+	stoppedCh := make(chan struct{})
```

Comment:
It's a bit confusing to have both `stopCh` and `stoppedCh`.  The two names are essentially the same; it would be better if the names made it clearer what the difference is.  Alternatively, we could use a `sync.WaitGroup` instead of `stoppedCh`.  Alternatively, we could just follow the simpler pattern in https://github.com/kubernetes/client-go/blob/master/examples/workqueue/main.go .  Note that the simpler pattern will pass the test func introduced in this PR.

Reply:
I renamed it to `workerDone`. I am also fine with using a `sync.WaitGroup`.  

To wait for the worker to complete seems to be a more graceful way for the `Run` function to exit. (this is not necessary to fix the deadlock bug though).

Reply:
It's not wrong for `Run` to wait for the worker goroutine to finish.  OTOH, it is not like other controllers.

Reply:
@MikeSpreitzer `Run` forks a long running go routine and I think it is responsible for waiting for it to finish, to allow graceful termination happen. The same rule should apply to all controllers if we care about graceful termination.

Since the primary focus of this PR is to fix the deadlock issue, I am going to remove the graceful bit from here and will follow up with a new PR for making it graceful (where we can have a larger discussion). I think for graceful termination of controllers to work end to end - starting with the primary go routine and then the post startup hooks to the controllers, every layer needs to play its part.